controller_manager:
  ros__parameters:
    use_sim_time: True
    update_rate: 100  # Hz
    ffw_swerve_drive_controller:
      type: ffw_swerve_drive_controller/SwerveDriveController

    # Joint State Broadcaster
    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster
      

ffw_swerve_drive_controller:
  ros__parameters:
    steering_joint_names: ["steering_front_left_joint", "steering_front_right_joint", "steering_rear_joint"]
    wheel_joint_names: ["wheel_front_left_joint", "wheel_front_right_joint", "wheel_rear_joint"]
    wheel_radius: 0.05 # [m]

    module_x_offsets: [ 0.2054,  0.2054, -0.2135]
    module_y_offsets: [ 0.2135, -0.2135,  0.0]
    module_angle_offsets: [ 0.0, 0.0, -0.0, -0.0] # [rad]
    module_steering_limit_lower: [ -1.570796, -1.570796, -1.570796, -1.570796 ] # [rad]
    module_steering_limit_upper: [  1.570796,  1.570796,  1.570796,  1.570796 ] # [rad]
    module_wheel_speed_limit_lower: [ -50.0, -50.0, -50.0, -50.0 ] # [rad/s]
    module_wheel_speed_limit_upper: [ 50.0, 50.0, 50.0, 50.0 ] # [rad/s]

    enabled_steering_flip: false # true or false
    steering_angular_velocity_limit: 1.0 # [rad/s]
    steering_alignment_angle_error_threshold: 0.10 # [rad]
    cmd_vel_timeout: 5.0  # [sec]

    # odometry
    odom_solver_method: "svd"  # "pseudo_inverse", "qr", "svd"
    odom_source: "feedback"  # "command" or "feedback", Choose if open-loop or not (feedback) is used for odometry calculation.
    base_frame_id: "base_link"
    odom_frame_id: "odom"
    enable_odom_tf: true
    pose_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.001]
    twist_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.001]
    velocity_rolling_window_size: 10

    # visualization
    enable_visualization: true
    visualization_marker_topic: "swerve_visualization_markers"
    visualization_update_time: 0.10 # [sec]

    # Velocity and acceleration limits
    # Whenever a min_* is unspecified, default to -max_*
    enabled_speed_limits: true
    publish_limited_velocity: true
    linear:
      x:
        has_velocity_limits    : true
        max_velocity           : 0.50  # m/s
        min_velocity           : -1.0 # m/s
        has_acceleration_limits: true
        max_acceleration       : 0.4  # m/s^2
        min_acceleration       : -1.0 # m/s^2
        has_jerk_limits        : true
        max_jerk               : 1.0  # m/s^3
        min_jerk               : -5.0  # m/s^3
      y:
        has_velocity_limits    : true
        max_velocity           : 0.50  # m/s
        min_velocity           : -1.0 # m/s
        has_acceleration_limits: true
        max_acceleration       : 0.4  # m/s^2
        min_acceleration       : -1.0 # m/s^2
        has_jerk_limits        : true
        max_jerk               : 1.0  # m/s^3
        min_jerk               : -5.0  # m/s^3
    angular:
      z:
        has_velocity_limits    : true
        max_velocity           : 1.7  # rad/s
        has_acceleration_limits: true
        max_acceleration       : 1.5  # rad/s^2
        has_jerk_limits        : true
        max_jerk               : 10.0  # rad/s^3
# Joint State Broadcaster
joint_state_broadcaster:
  ros__parameters:
    {}
